vehicle_gateway_demo/src/follower.cpp

```diff
@@ -0,0 +1,171 @@
+// Copyright 2023 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <fstream>
+#include <iostream>
+#include <string>
+
+#include <nlohmann/json.hpp>
+#include <pluginlib/class_loader.hpp>
+#include <rclcpp/rclcpp.hpp>
+#include <vehicle_gateway_px4/vehicle_gateway_px4.hpp>
+#ifdef _WIN32
+#include <Windows.h>
+#else
+#include <unistd.h>
+#endif
+
+using nlohmann::json;
+using std::string;
+
+class Follower
+{
+public:
+  Follower(const int vehicle_id)
+  {
+    this->loader_ = std::make_shared<pluginlib::ClassLoader<vehicle_gateway::VehicleGateway>>(
+      "vehicle_gateway", "vehicle_gateway::VehicleGateway");
+    this->gateway_ = this->loader_->createSharedInstance("vehicle_gateway_px4::VehicleGatewayPX4");
+    this->gateway_->set_vehicle_id(vehicle_id);
+    this->gateway_->init(0, nullptr);
+  }
+
+  void destroy()
+  {
+    if (this->gateway_) {
+      this->gateway_->destroy();
+      this->gateway_ = nullptr;
+    }
+    if (this->loader_) {
+      this->loader_->unloadLibraryForClass(
+        "vehicle_gateway_px4::VehicleGatewayPX4");
+      this->loader_ = nullptr;
+    }
+  }
+
+  std::shared_ptr<vehicle_gateway::VehicleGateway> gateway_;
+  void state_handler(const z_sample_t * sample);
+
+private:
+  std::shared_ptr<pluginlib::ClassLoader<vehicle_gateway::VehicleGateway>> loader_;
+  bool following_active{false};
+};
+
+void Follower::state_handler(const z_sample_t * sample)
+{
+  json j;
+  try {
+    j = json::parse(
+      std::string(
+        reinterpret_cast<const char *>(sample->payload.start),
+        sample->payload.len));
+  } catch (...) {
+    printf("error parsing json\n");
+    return;
+  }
+  // todo: test for required fields
+  const int sender_id = j["id"].get<int>();
+  // std::cout << "sender_id " << sender_id << '\n';
+  if (sender_id != 1) {
+    // printf("ignoring non-leader state (id %d)\n", sender_id);
+    return;
+  }
+  const double leader_north = j["north"].get<double>();
+  const double leader_east = j["east"].get<double>();
+  const double leader_down = j["down"].get<double>();
+  const double airspeed = j["airspeed"].get<double>();
+  printf(
+    "leader position: (%.3f, %.3f, %.3f)\n",
+    leader_north, leader_east, leader_down);
+
+  if (!this->following_active) {
+    this->start = std::chrono::steady_clock::now();
```

Comment:
Is `start` declared anywhere, or am I missing something?

Reply:
After removing this line, things built for me.

Reply:
https://github.com/osrf/vehicle_gateway/pull/122/commits/bdd2b71b3d3a46b8d52f9b080e542ffd5c4597cc

Reply:
I see it was removed in bdd2b71b3d3a46b8d52f9b080e542ffd5c4597cc